curve25519-dalek/src/hazmat/lazy_field25519.rs

```diff
@@ -0,0 +1,130 @@
+            + Add<
+                CapacityUsed,
+                Output: Unsigned + IsLessOrEqual<<Underlying as UnderlyingCapacity>::Capacity>,
+            >,
+        T: LazyField<V, Underlying = Self::Underlying>,
+    >(
+        self,
+        other: &T,
+    ) -> impl LazyField<
+        <V as Add<CapacityUsed>>::Output,
+        Capacity = Self::Capacity,
+        Underlying = Self::Underlying,
+        Output = <Self as Reducible>::Output,
+    > {
+        FieldElement::<<V as Add<CapacityUsed>>::Output>::from(&self.0.0 + &other.as_underlying().0)
+    }
+
+    fn mul<V: Unsigned, T: LazyField<V, Underlying = Self::Underlying>>(
+        self,
+        other: &T,
+    ) -> <Self as Reducible>::Output {
+        let unreduced = &self.0.0 * &other.as_underlying().0;
+        FieldElement::from(Underlying::from_bytes(&unreduced.to_bytes()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::hazmat::lazy_field::{EagerField, LazyField, LazyFieldWithCapacity, Reducible};
+    use typenum::{B1, U2, U3, type_operators::IsLessOrEqual};
+
+    fn add_triple_then_mul<F: LazyFieldWithCapacity<U3>>(
+        a: F,
+        b: F,
+        c: F,
+        d: F,
+        e: F,
+        f: F,
+    ) -> <F as Reducible>::Output
+    where
+        U2: IsLessOrEqual<F::Capacity, Output = B1>,
+        U3: IsLessOrEqual<F::Capacity, Output = B1>,
+    {
+        let ab = a.add(&b);
+        let abc = ab.add(&c);
+        let de = d.add(&e);
+        let def = de.add(&f);
+        abc.mul(&def)
+    }
+
+    #[test]
+    fn lazy_add_then_mul() {
+        use crate::hazmat::FieldElement;
+        use core::marker::PhantomData;
+        use ff::Field;
+        use rand_core::{OsRng, TryRngCore};
+
+        let mut rng = OsRng.unwrap_err();
+
+        let a = FieldElement::random(&mut rng);
+        let b = FieldElement::random(&mut rng);
+        let c = FieldElement::random(&mut rng);
+        let d = FieldElement::random(&mut rng);
+        let e = FieldElement::random(&mut rng);
+        let f = FieldElement::random(&mut rng);
+        let expected = (a + b + c) * (d + e + f);
+
+        assert_eq!(
+            LazyField::add(a, &b)
+                .add(&c)
+                .mul(&LazyField::add(d, &e).add(&f)),
+            expected
+        );
+        assert_eq!(add_triple_then_mul(a, b, c, d, e, f), expected);
+
+        let a = EagerField(a, PhantomData::<typenum::U1>);
+        let b = EagerField(b, PhantomData::<typenum::U1>);
+        let c = EagerField(c, PhantomData::<typenum::U1>);
+        let d = EagerField(d, PhantomData::<typenum::U1>);
+        let e = EagerField(e, PhantomData::<typenum::U1>);
+        let f = EagerField(f, PhantomData::<typenum::U1>);
+
+        assert_eq!(
+            LazyField::add(a, &b)
+                .add(&c)
+                .mul(&LazyField::add(d, &e).add(&f))
+                .0,
+            expected
+        );
+        assert_eq!(add_triple_then_mul(a, b, c, d, e, f).0, expected);
+    }
```

Comment:
If you used `proptests` for these, we could get the information needed to reproduce CI failures like: https://github.com/dalek-cryptography/curve25519-dalek/actions/runs/17444578081/job/49535734542?pr=816#step:8:205